### PR TITLE
feat: merge scope where conditions using the and operator

### PR DESCRIPTION
**Pull request by EtienneTurc**
### Pull Request Checklist

- [x] Have you added new tests to prevent regressions?
- [x] Does `yarn test` or `yarn test-DIALECT` pass with this change (including linting)?
- [x] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [x] Did you update the typescript typings accordingly (if applicable)?
- [x] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Did you follow the commit message conventions explained in [CONTRIBUTING.md](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md)?

### Description Of Change

Issues related:
- https://github.com/sequelize/sequelize/issues/11588
- https://github.com/sequelize/sequelize/issues/10873

Pull request that tried to solve these issues:
- https://github.com/sequelize/sequelize/pull/11805

This pr adds the following elements:
- A way to merge scopes with where clauses using the `AND` operator instead of overwriting it with the last value. It supports merging attributes, `[Op.and]` and `[Op.or]`. 
- The flag `mergeWhereScopesWithAndOperator` to enable this behavior for a model or all the models.

Examples:
```js
// Attributes merging

// Scope 1
whereAttributeIs1: {
  where: {
    field: 1,
  },
},
// Scope 2
whereAttributeIs2: {
  where: {
    field: 2,
  },
}

// Current behavior (without mergeWhereScopesWithAndOperator)
myModel.scope('whereAttributeIs1', 'whereAttributeIs2') ._scope ===
{
  where: {
    field: 2,
  }
}

// Expected behavior (with mergeWhereScopesWithAndOperator)
myModel.scope('whereAttributeIs1', 'whereAttributeIs2') ._scope ===
{
  where: {
    field: {
      [Op.and]: [1, 2]
    }
  }
}
```

```js
// Op.and merging

// Scope 1
whereOpAnd1: {
  where: {
    [Op.and]: [{ field: 1 }],
  },
},
// Scope 2
whereOpAnd2: {
  where: {
    [Op.and]: [{ field: 2 }],
  },
},

// Current behavior (without mergeWhereScopesWithAndOperator)
myModel.scope('whereOpAnd1', 'whereOpAnd2') ._scope ===
{
  where: {
    [Op.and]: [{ field: 2 }],
  },
}

// Expected behavior (with mergeWhereScopesWithAndOperator)
myModel.scope('whereOpAnd1', 'whereOpAnd2') ._scope ===
{
  where: {
    [Op.and]: [{ field: 1 }, { field: 2 }],
  },
}
```

```js
// Op.or merging

// Scope 1
whereOpOr1: {
  where: {
    [Op.or]: [{ field: 1 }],
  },
},
// Scope 2
whereOpOr2: {
  where: {
    [Op.or]: [{ field: 2 }],
  },
},

// Current behavior (without mergeWhereScopesWithAndOperator)
myModel.scope('whereOpOr1', 'whereOpOr2') ._scope ===
{
  where: {
    [Op.or]: [{ field: 2 }]
  },
}

// Expected behavior (with mergeWhereScopesWithAndOperator)
myModel.scope('whereOpAnd1', 'whereOpAnd2') ._scope ===
{
  where: {
    [Op.and]: [
      { [Op.or]: [{ field: 1 }] }, 
      { [Op.or]: [{ field: 2 }] }
    ],
  },
}
```

Why should this be in Sequelize ?

The current behavior for merging different attributes already acts as an `AND` operator. Thus it is natural to expect the  
same behavior for different conditions on the same attribute.

### CODE REVIEW

**Comment by ephys** [comment]
Wouldn't it be easier and less error prone to simply combine where conditions with `Op.and`?

So:

```typescript
// Scope 1
whereAttributeIs1: {
  where: { field: 1 },
},
// Scope 2
whereAttributeIs2: {
  where: { field: 2 },
}

myModel.scope('whereAttributeIs1', 'whereAttributeIs2') ._scope === 
{
  where: {
    [Op.and]: [
      { field: 1 },
      { field: 2 },
    ]
  }
}
```

**Comment by EtienneTurc** [comment]
> bl0cknumber

Thanks for the reply and the suggestion. Indeed, it would be nice to update the manual documentation with this new feature. I assume it would be best to add a new sub section in the merging section in https://sequelize.org/v6/manual/scopes.html (https://github.com/EtienneTurc/sequelize/blob/add-option-to-merge-where-scope-with-and-operator/docs/manual/other-topics/scopes.md).

> ephys

Thanks for the reply and the suggestion. Indeed, it looks easier and thus the previous examples will look like these:

```js
// Attributes merging

// Scope 1
whereAttributeIs1: {
  where: {
    field: 1,
  },
},
// Scope 2
whereAttributeIs2: {
  where: {
    field: 2,
  },
}

// New expected behavior
myModel.scope('whereAttributeIs1', 'whereAttributeIs2')._scope ===
{
  where: {
    [Op.and]: [
      {field: 1},
      {field: 2}
   ]
  }
}
```

```js
// Op.and merging

// Scope 1
whereOpAnd1: {
  where: {
    [Op.and]: [{ field: 1 }],
  },
},
// Scope 2
whereOpAnd2: {
  where: {
    [Op.and]: [{ field: 2 }],
  },
},

// New expected behavior
myModel.scope('whereOpAnd1', 'whereOpAnd2')._scope ===
{
  where: {
    [Op.and]: [
      { field: 1 },
      { field: 2 }
    ],
  },
}
```

```js
// Op.or merging

// Scope 1
whereOpOr1: {
  where: {
    [Op.or]: [{ field: 1 }],
  },
},
// Scope 2
whereOpOr2: {
  where: {
    [Op.or]: [{ field: 2 }],
  },
},

// New expected behavior
myModel.scope('whereOpAnd1', 'whereOpAnd2')._scope ===
{
  where: {
    [Op.and]: [
      { [Op.or]: [{ field: 1 }] },
      { [Op.or]: [{ field: 2 }] }
    ],
  },
}
```

But let's also note that this implementation will also impact those cases (even if this should not be an issue):
```js
// Scope 1
whereAttributeIs1: {
  where: {
    field1: 1,
  },
},
// Scope 2
whereAttributeIs2: {
  where: {
    field2: 2,
  },
}

// Current behavior
myModel.scope('whereAttributeIs1', 'whereAttributeIs2')._scope ===
{
  where: {
    field1: 1,
    field2: 2
  },
}

// New expected behavior
myModel.scope('whereAttributeIs1', 'whereAttributeIs2')._scope ===
{
  where: {
    [Op.and]: [
      {field1: 1},
      {field2: 2}
   ]
  }
}
```

(Currently making the modifications)

**Comment by EtienneTurc** [comment]
I've made the requested changes. Note that it slightly changes the behavior of merging `Op.and`:

```js
// Op.and merging

// Scope 1
whereOpAnd1: {
  where: {
    [Op.and]: [{ field: 1 }],
  },
},
// Scope 2
whereOpAnd2: {
  where: {
    [Op.and]: [{ field: 2 }],
  },
},

// New behavior
myModel.scope('whereOpAnd1', 'whereOpAnd2')._scope ===
{
  where: {
    [Op.and]: [
      { [Op.and]: [{ field: 1  }] },
      { [Op.and]: [{ field: 2  }] }
    ],
  },
}
```

It drastically simplifies the code and makes debugging easier because where clauses from different scopes are not mixed together.

I am waiting for you to approve these changes before adding a new sub-section in the manual documentation :)

**Review by ephys — CHANGES_REQUESTED** [review]
From what I can see, `mergeWhereScopesWithAndOperator` is in the `Model`'s options. I think that's fine because IMO the current merge behavior is just plain wrong and should be replaced with the new one.

---

This is targeting `main` (v7 alpha). I think it would be better to only introduce this flag in v6 to avoid a breaking change, and use the new behavior as the default behavior in v7.

That would mean:
- Open a new PR that targets the v6 branch and copy-paste the current code there.
- Change this PR to make merging with AND the only behavior in v7, without a flag

---

Can you add a few tests that verify that `Model.findAll` merges the scopes using the right strategy?

**Comment by EtienneTurc** [comment]
Thanks for the review. I also believe that this behavior should be the default one, so I am glad that you are willing to use it for the v7.

I will make the two PRs in the following days :)

**Comment by EtienneTurc** [comment]
I've made the requested modifications. So now this pr (for the v7) changes the merging behavior from the `overwrite` one to the `and` one. Thus this change is breaking and tests have been modified/added to take this into consideration.

For the v6, I've opened this pr:
- https://github.com/sequelize/sequelize/pull/14152

NB: I took the liberty to name the current behavior `overwrite` instead of `assign` as suggested; I find it more explicit.

**Comment by ephys** [comment]
Thanks! I'll try to not take too long before reviewing both PRs.

I'm not the most knowledgeable users of scopes (never used them, actually) so I'd like another member of @sequelize/maintainers to review this too, but based on my understanding of scopes, these changes look good.

In-depth review as soon as I can

**Comment by EtienneTurc** [comment]
I've made the requested changes (on both prs). 

I was reluctant at first to unpack `Op.and` because in some cases it mixes multiples scopes together making it a bit harder to debug (but still functional). However, in most cases, this operand is just unnecessary, so I guess it makes sense to remove it.

File: test/unit/model/include.test.js
Comment (ephys) [inline test/unit/model/include.test.js:183]: You reverted your usage of deepEqual, is that intended? Didn't you add it to be able to compare `symbol`?

I saw a commit fail because it was trying to compare 'object with null prototype' to 'object', I assume that's why you reverted?

https://github.com/chaijs/chai/issues/1054 has had some movement but I don't think they'll release the change until chai 5

---

also if `deep.equals` and `deep.equal` are the same, can you revert to `deep.equals` to reduce risk of conflict?

File: test/unit/model/include.test.js
Comment (ephys) [inline test/unit/model/include.test.js:212]: Aah I see you test it manually, then you can ignore the first part of my previous comment :)

Can you add `// TODO [chai>5]: simplify once '.deep.equal' includes support for symbols (https://github.com/chaijs/chai/issues/1054)`

**Review by ephys — CHANGES_REQUESTED** [review]
Looking good :)
I agree with your concern, if we didn't use _.merge we could have enough information to only add [Op.and] when necessary without unpacking existing [Op.and] but I think this is good enough

Only 3 nits and it's good on my end
The concept was also approved by other team members

**Comment by EtienneTurc** [comment]
> Looking good :) 

Great ! I have done the few changes.

File: test/unit/model/include.test.js
Comment (ephys) [inline test/unit/model/include.test.js:182]:
```suggestion
        expect(options.include[0]).to.have.property('where').which.deep.equals({ active: true });
```

not sure why the linter didn't catch this
will merge as soon as this suggestion is committed :)

**Review by ephys — APPROVED** [review]

File: test/unit/model/include.test.js
Comment (EtienneTurc) [reply test/unit/model/include.test.js]: Don't know either 🤔

Comment (EtienneTurc) [reply test/unit/model/include.test.js]: I've removed the extra space, so it is good for the merge ;)

**Review by ephys — APPROVED** [review]

**Comment by ephys** [comment]
And here we go! 🎊

**Pull request merged**
